### PR TITLE
Use prior phylanx defs

### PR DESCRIPTION
**Pull request by stevenrbrandt**
This change makes it possible to use one Phylanx defined function inside another. This is especially important for Jupyter.

### CODE REVIEW

File: python/src/bindings/binding_helpers.hpp [hunk 1]
```diff
@@ -117,16 +117,24 @@ namespace phylanx { namespace bindings
...
     {
         namespace et = phylanx::execution_tree;
         return hpx::threads::run_as_hpx_thread(
             [&]() -> et::primitive_argument_type
             {
                 try
                 {
-                    et::compiler::function_list eval_snippets;
-                    auto x = et::compile(xexpr, eval_snippets);
+                    static phylanx::execution_tree::compiler::environment *eval_env = nullptr;
```
Comment (hkaiser) [inline python/src/bindings/binding_helpers.hpp]: Why pointers? Where is the allocated memory free'd?

File: python/phylanx/ast/transformation.py [hunk 2]
```diff
@@ -7,13 +7,45 @@
...
+
+
+class DoNothing:
+    def __enter__(self):
+        pass
+
+    def __exit__(self, type, value, traceback):
+        pass
+
+
```
Comment (hkaiser) [inline python/phylanx/ast/transformation.py]: Is this change related?

**Comment by hkaiser** [comment]
Could you please add a test for this new functionality?

**Comment by hkaiser** [comment]
@stevenrbrandt are we sure that we will not need to separate several different 'translation units'? The way this is written now everything is dumped into the same 'namespace'. 

I think we should make this more modular, for instance compiling each Python module into it's own PhySL environment.

File: python/src/bindings/binding_helpers.hpp (hunk 1 above)
Comment (stevenrbrandt) [reply python/src/bindings/binding_helpers.hpp]: It is not freed, of course. The plan was for a single, long-running set of defs to be available globally. My experience is that static objects don't initialize properly, so I use pointers to objects. I also tried making an extern version of these two objects, but this also caused memory problems.

Comment (hkaiser) [reply python/src/bindings/binding_helpers.hpp]: Static objects initialize just fine, starting with C++11 their initialization is thread-safe even (which your solution is not, btw).

**Comment by stevenrbrandt** [comment]
I've refactored this to have a compiler_state object that can be created in Python and passed around. It works, except the following problem occurs at shutdown:
```
terminate called after throwing an instance of 'std::invalid_argument'
  what():  hpx::resource::get_partitioner() can be called only after the resource partitioner has been allowed to parse the command line options.
Aborted (core dumped)
```

**Comment by hkaiser** [comment]
@stevenrbrandt how do I reproduce this error during shutdown?

**Comment by hkaiser** [comment]
@stevenrbrandt this happens during destruction of the (global) `compiler_state` object during module unload. This happens after HPX has been released already. I think we'll need to add a weak reference to the phylanx extension library for each `compiler_state` instance created. Let me see what I can do.

**Comment by hkaiser** [comment]
@stevenrbrandt things should be ok now.

File: python/phylanx/ast/transformation.py [hunk 3]
```diff
@@ -442,15 +446,14 @@ def __init__(self, f):
             assert len(tree.body) == 1
 
             if target == "PhySL":
-                self.__physl_src__ = '%s(%s)\n' % (
-                    full_node_name(tree.body[0], 'block'),
-                    transformation.recompile(tree))
+                self.__physl_src__ = transformation.recompile(tree)
+                et.eval(self.__physl_src__, self.cs)
```
Comment (hkaiser) [inline python/phylanx/ast/transformation.py]: Should things be evaluated whenever the decorators is created?

Comment (stevenrbrandt) [reply python/phylanx/ast/transformation.py]: @hkaiser It seems to work now. Yes, when the decorator is evaluated, the function is compiled. This is an improvement over what happened before, where the function was compiled on each evaluation. When the function is called, eval is called again, but instead of the function definition, only the function name is needed.

Comment (hkaiser) [reply python/phylanx/ast/transformation.py]: Neat trick!

Comment (hkaiser) [reply python/phylanx/ast/transformation.py]: OTOH, doesn't this also mean that the function is evaluated twice? Once without arguments (which could btw fail) and once with them? It might be better to explicitly separate compilation and evaluation in the Python interface.

**Comment by hkaiser** [comment]
@stevenrbrandt could you address the failing builds for us to be able to go ahead with merging, please?
